### PR TITLE
Mac os support

### PR DESCRIPTION
**Pull request by woutersamaey**
Should work for most Mac OS users. Limited testing on a virtualized fresh Mac OS Mojave install, more testing on my already setup machine.

Unfortunatly testing is hard since Docker won't start on a virtualized Mac, but regardless, this should just work!

Since Mac OS users are not accustomed to "root" or "sudo", everything works using the normal user. Only when installing homebrew (package manager used to install Docker and Docker compose) you will be prompted for your password.

Would like to get some more feedback on the first run for fresh installs (people who don't have docker of homebrew yet), but should be 95% fine.

All scripts work (up, down, install, upgrade).

The code in btcpay-setup.sh is getting harder to read, but did my best to apply more structure and comments.

### CODE REVIEW

File: btcpay-setup.sh [hunk 1]
```diff
@@ -383,25 +448,33 @@ stop on runlevel [!2345]
...
+
 cd "$(dirname $BTCPAY_ENV_FILE)"
 
-if [ ! -z "$OLD_BTCPAY_DOCKER_COMPOSE" ] && [ "$OLD_BTCPAY_DOCKER_COMPOSE" != "$BTCPAY_DOCKER_COMPOSE" ]; then
+if [[ ! -z "$OLD_BTCPAY_DOCKER_COMPOSE" ]] && [[ "$OLD_BTCPAY_DOCKER_COMPOSE" != "$BTCPAY_DOCKER_COMPOSE" ]]; then
     echo "Closing old docker-compose at $OLD_BTCPAY_DOCKER_COMPOSE..."
     docker-compose -f "$OLD_BTCPAY_DOCKER_COMPOSE" down -t "${COMPOSE_HTTP_TIMEOUT:-180}"
 fi
 
+return
```
Comment (NicolasDorier) [inline btcpay-setup.sh]: remove this

File: btcpay-setup.sh [hunk 2]
```diff
@@ -327,50 +388,54 @@ if [ "$BTCPAYGEN_OLD_PREGEN" == "true" ]; then
...
+if [[ -x "$(command -v systemctl)" ]]; then
+	# Use systemd
+	if [[ -e "/etc/init/start_containers.conf" ]]; then
+		echo -e "Uninstalling upstart script /etc/init/start_containers.conf"
+		rm "/etc/init/start_containers.conf"
+		initctl reload-configuration
+	fi
+	echo "Adding btcpayserver.service to systemd"
+	echo "
+	[Unit]
```
Comment (NicolasDorier) [inline btcpay-setup.sh]: don't change the indentation here, this will change indentation in the systemctl.

Comment (woutersamaey) [reply btcpay-setup.sh]: You're right. Fixed it!

File: btcpay-setup.sh (hunk 1 above)
Comment (woutersamaey) [reply btcpay-setup.sh]: Must've left this during testing. Now gone.

**Comment by woutersamaey** [comment]
If you prefer, I can squash the commits into 1

File: btcpay-setup.sh [hunk 3]
```diff
@@ -1,22 +1,49 @@
...
+	if [[ $EUID -eq 0 ]]; then
+		# Running as root is discouraged on Mac OS. Run under the current user instead.
+        echo "This script should not be run as root."
+        return
+    fi
+
+	BASH_PROFILE_SCRIPT="$HOME/btcpay-env.sh"
+
+	# Mac OS doesn't use /etc/profile.d/xxx.sh. Instead we create a new file and load that from ~/.bash_profile
+	if [ ! -f "$FILE" ]; then
```
Comment (NicolasDorier) [inline btcpay-setup.sh]: what is `$FILE`?

Comment (woutersamaey) [reply btcpay-setup.sh]: Woops. Something must've gone wrong with stashing my changes and moving files back and forth between my test VM. 

Fixed it

File: btcpay-setup.sh [hunk 4]
```diff
@@ -202,19 +232,23 @@ BTCPAY_ANNOUNCEABLE_HOST:$BTCPAY_ANNOUNCEABLE_HOST
...
     echo "NBITCOIN_NETWORK should be equal to mainnet, testnet or regtest"
 fi
 
-# Put the variables in /etc/profile.d when a user log interactively
-touch "/etc/profile.d/btcpay-env.sh"
-echo "#!/bin/bash
 
+
+# Init the variables when a user log interactively
+touch ${BASH_PROFILE_SCRIPT}
```
Comment (NicolasDorier) [inline btcpay-setup.sh]: duplicated

File: btcpay-setup.sh [hunk 5]
```diff
@@ -239,20 +273,21 @@ if cat \"\$BTCPAY_ENV_FILE\" &> /dev/null; then
...
-echo -e "BTCPay Server environment variables successfully saved in /etc/profile.d/btcpay-env.sh\n"
+echo -e "BTCPay Server environment variables successfully saved in $BASH_PROFILE_SCRIPT\n"
 
 
 btcpay_update_docker_env
 
 echo -e "BTCPay Server docker-compose parameters saved in $BTCPAY_ENV_FILE\n"
 
-. /etc/profile.d/btcpay-env.sh
+. ${BASH_PROFILE_SCRIPT}
```
Comment (NicolasDorier) [inline btcpay-setup.sh]: `"${BASH_PROFILE_SCRIPT}"`

File: btcpay-setup.sh (hunk 4 above)
Comment (NicolasDorier) [reply btcpay-setup.sh]: `"${BASH_PROFILE_SCRIPT}"`

File: btcpay-restart.sh [hunk 6]
```diff
@@ -1,6 +1,15 @@
...
+if [[ "$OSTYPE" == "darwin"* ]]; then
+	# Mac OS
+	BASH_PROFILE_SCRIPT="$HOME/btcpay-env.sh"
+
+else
+	# Linux
+	BASH_PROFILE_SCRIPT="/etc/profile.d/btcpay-env.sh"
+fi
+
+. ${BASH_PROFILE_SCRIPT}
```
Comment (NicolasDorier) [inline btcpay-restart.sh]: `add "`

File: btcpay-down.sh [hunk 7]
```diff
@@ -1,6 +1,16 @@
...
+if [[ "$OSTYPE" == "darwin"* ]]; then
+	# Mac OS
+	BASH_PROFILE_SCRIPT="$HOME/btcpay-env.sh"
+
+else
+	# Linux
+	BASH_PROFILE_SCRIPT="/etc/profile.d/btcpay-env.sh"
+fi
+
+. ${BASH_PROFILE_SCRIPT}
```
Comment (NicolasDorier) [inline btcpay-down.sh]: add `"`

Comment (woutersamaey) [reply btcpay-down.sh]: You mean `. ${BASH_PROFILE_SCRIPT}` should become `. "$BASH_PROFILE_SCRIPT"` ?
Is this causing issues for you? Syntax looks OK to me.

File: btcpay-restart.sh (hunk 6 above)
Comment (woutersamaey) [reply btcpay-restart.sh]: Duplicate, see below.

File: btcpay-down.sh (hunk 7 above)
Comment (NicolasDorier) [reply btcpay-down.sh]: @woutersamaey well with `. ${BASH_PROFILE_SCRIPT}` if there is a space in `$BASH_PROFILE_SCRIPT` this would fail right?

Comment (woutersamaey) [reply btcpay-down.sh]: Right. Figured this out on my own and already commited new changes.

File: btcpay-update.sh [hunk 8]
```diff
@@ -22,21 +31,28 @@ if [[ "$1" != "--skip-git-pull" ]]; then
...
+
 . ./build.sh
 if [ "$BTCPAYGEN_OLD_PREGEN" == "true" ]; then
     cp Generated/docker-compose.generated.yml $BTCPAY_DOCKER_COMPOSE
 fi
 
-if ! grep -Fxq "export COMPOSE_HTTP_TIMEOUT=\"180\"" "/etc/profile.d/btcpay-env.sh"; then
-    echo "export COMPOSE_HTTP_TIMEOUT=\"180\"" >> /etc/profile.d/btcpay-env.sh
+if ! grep -Fxq "export COMPOSE_HTTP_TIMEOUT=\"180\"" "$BASH_PROFILE_SCRIPT"; then
+    echo "export COMPOSE_HTTP_TIMEOUT=\"180\"" >> ${BASH_PROFILE_SCRIPT}
```
Comment (NicolasDorier) [inline btcpay-update.sh]: need quote

File: btcpay-setup.sh [hunk 9]
```diff
@@ -383,25 +448,32 @@ stop on runlevel [!2345]
...
-    echo "BTCPay Server started"
-fi
+
+	if $START; then
+		initctl reload-configuration
+		echo "BTCPay Server started"
+	fi
 fi
 
+#if [[ "$OSTYPE" == "darwin"* ]]; then
```
Comment (NicolasDorier) [inline btcpay-setup.sh]: delete itif not used

Comment (woutersamaey) [reply btcpay-setup.sh]: Should I keep the TODO? Not sure why the upstart script exists. Can't Docker start the containers on its own without the upstart script?

Comment (NicolasDorier) [reply btcpay-setup.sh]: While docker start automatically, I don't think it starts a docker-compose service automatically.

File: btcpay-setup.sh [hunk 10]
```diff
@@ -344,33 +406,36 @@ Requires=docker.service network-online.target
...
-systemctl enable btcpayserver
-if $START; then
-    echo "BTCPay Server starting... this can take 5 to 10 minutes..."
-    systemctl start btcpayserver
-    echo "BTCPay Server started"
-fi
-else # Use upstart
-echo "Using upstart"
-echo "
+if [[ -x "$(command -v initctl)" ]]; then
```
[the remark this answers is not in the document]
Comment (woutersamaey) [reply btcpay-setup.sh]: Turned this into an "elif". Just "else" is no good for Mac.

File: btcpay-down.sh (hunk 7 above)
Comment (NicolasDorier) [reply btcpay-down.sh]: you did not fixed this one btw!

Comment (woutersamaey) [reply btcpay-down.sh]: Added the quotes now.

File: btcpay-setup.sh (hunk 9 above)
Comment (woutersamaey) [reply btcpay-setup.sh]: Just did some more testing and can confirm that the containers go back up after a reboot.

**Comment by woutersamaey** [comment]
Just did some final testing and can confirm that all the commands are working, including after rebooting. The containers are there. Everything seams 100% functional now on Mac OS X Mojave (and probably all other Mac OS versions that can run docker)

@NicolasDorier any remaining issues you detect?

**Comment by woutersamaey** [comment]
I don't expect many Mac users will run this in production, but it's a great way to experiment and try BTCPay Server for Mac users.
Kinda makes me think of Pierre Rochard's Node Launcher, which is also onboarding people.

**Comment by NicolasDorier** [comment]
@woutersamaey I think it is good will do final review.

I am happy to see btcpayserver run in as much environment as I can.

**Comment by woutersamaey** [comment]
Still testing @NicolasDorier ?

**Pull request merged**